packages/alfa-style/src/property/background.ts

```diff
@@ -70,45 +126,712 @@ export namespace Background {
+        | Keyword<"padding-box">
+        | Keyword<"content-box">;
+    }
+
+    export type Computed = Specified;
+  }
+
+  /**
+   * @see https://drafts.csswg.org/css-backgrounds/#propdef-background-size
+   */
+
+  const parseSize = either(
+    pair(
+      either(Length.parse, Keyword.parse("auto")),
+      map(option(either(Length.parse, Keyword.parse("auto"))), (y) =>
+        y.getOrElse(() => Keyword.of("auto"))
+      )
     ),
+    Keyword.parse("contain", "cover")
+  );
+
+  const parseSizeList = map(
+    separatedList(
+      parseSize,
+      delimited(option(Token.parseWhitespace), Token.parseComma)
+    ),
+    (sizes) => List.of(sizes, ", ")
+  );
+
+  export const Size: Property<Size.Specified, Size.Computed> = Property.of(
+    List.of([[Keyword.of("auto"), Keyword.of("auto")]], ", "),
+    parseSizeList,
     (style) =>
-      style.specified("background-image").map((images) =>
+      style.specified("background-size").map((sizes) =>
         List.of(
-          Iterable.map(images, (image) => {
-            switch (image.type) {
-              case "keyword":
-                return image;
-
-              case "image":
-                return resolveImage(image, style);
+          Iterable.map(sizes, (size) => {
+            if (Keyword.isKeyword(size)) {
+              return size;
             }
+
+            const [x, y] = size;
+
+            return [
+              x.type === "length" ? Resolver.length(x, style) : x,
+              y.type === "length" ? Resolver.length(y, style) : y,
+            ];
           }),
           ", "
         )
       )
   );
 
+  export namespace Size {
+    export type Contain = Keyword<"contain">;
+
+    export type Cover = Keyword<"cover">;
+
+    export type Auto = Keyword<"auto">;
+
+    export type Specified = List<Specified.Item>;
+
+    export namespace Specified {
+      export type Item =
+        | [Length | Percentage | Auto, Length | Percentage | Auto]
+        | Cover
+        | Contain;
+    }
+
+    export type Computed = List<
+      | [Length<"px"> | Percentage | Auto, Length<"px"> | Percentage | Auto]
+      | Cover
+      | Contain
+    >;
+  }
+
+  const parseBackgroundLayer = map(
+    any<
+      Slice<Token>,
+      Array<
+        | ["background-color", Color.Specified]
+        | ["background-image", Image.Specified.Item]
+        | ["background-position-x", Position.X.Specified.Item]
+        | ["background-position-y", Position.Y.Specified.Item]
+        | ["background-size", Size.Specified.Item]
+        | ["background-repeat-x", Repeat.X.Specified.Item]
+        | ["background-repeat-y", Repeat.Y.Specified.Item]
+        | ["background-attachment", Attachment.Specified.Item]
+        | ["background-origin", Origin.Specified.Item]
+        | ["background-clip", Clip.Specified.Item]
+      >,
+      string
+    >(
+      map(delimited(option(Token.parseWhitespace), parseColor), (color) => [
+        ["background-color", color],
+      ]),
```

Comment:
I think `parseBackgroundLayer` would be more readable by having the various components of `any` be extracted and named individually. That function is huge and reading parser combinators is not easy…

Reply:
I've added some comments that names the different components. The issue with the individual parsers out of the background layer parser is that they're really specialised versions of the individual component parsers. For example, `parseColor` already exists and calling the one used the background layer parser `parseColorInBackgroundLayer` or something similar seems like a bad idea.

Reply:
Ah, yes. Now I see the structure of that parser a bit better 🙈 It's already just a bunch of wrappers around existing parser to add the `[key, value]` structure that is needed…
I guess the comments help 💯